### PR TITLE
C executable stanza [WIP]

### PR DESCRIPTION
**Pull request by rgrinberg**
Pushing this as WIP as there are some questions regarding the handling of source files that I'd like to answer first.

* Should {c,cxx}_names be required in the C_executables stanza? This seems to be symmetric to the library stanza but on the other hand, we tend to infer the sources for a stanza.

* Should we tweak `Dir_contents.t` to handle C sources as well. We touched that point of friction with coq once already. Seems like we'll need to address it here as well.

### CODE REVIEW

**Comment by emillon** [comment]
For `c_names`, I think that OSL would work: by default, every C file is a C name, but this can also be overridden.

How are dependencies computed? Every C file depends on all H files, or something like that?

**Comment by rgrinberg** [comment]
> For c_names, I think that OSL would work: by default, every C file is a C name, but this can also be overridden.

Yeah, I'll try to make this work but unfortunately there's no machinery for this stuff for c stubs.

> How are dependencies computed? Every C file depends on all H files, or something like that?

That's the plan.

**Comment by emillon** [comment]
Sounds good to me!

**Comment by rgrinberg** [comment]
> BTW, when we compile C stuff using ocamlc, we get a bunch of extra flags that are meant to build and link C code using the OCaml runtime. Should we have an option to use the plain C compiler without extra flags here? That might be useful in some situations.

My plan was to use the plain C compiler for executables. If the user would like to add those extra flags, then they may do so by substitution a variable.

**Comment by avsm** [comment]
The plain C compiler would indeed be appreciated for this, since ocamlc would force the additional of the caml include library (which is problematic for some test cases that would use this stanza I suspect -- similar header files)

File: src/dune_file.ml [hunk 1]
```diff
@@ -1930,10 +1930,14 @@ module Stanzas = struct
...
     ; "executables", Executables.multi >>| execs (fun e -> Executables e)
-    ; "c_executable"
-    , C_executables.single >>| execs (fun e -> C_executables e)
-    ; "c_executables",
-      C_executables.multi >>| execs (fun e -> C_executables e)
+    ; "c_executable",
+      (let%map () = Syntax.since Stanza.syntax (1, 7)
+       and t = C_executables.single in
+       execs (fun e -> C_executables e) t)
+    ; "c_executable",
```
Comment (xclerc) [inline src/dune_file.ml]: Shouldn't this be `"c_executables"`?

Comment (rgrinberg) [reply src/dune_file.ml]: Thanks. fixed.

**Comment by rgrinberg** [comment]
As I'm implementing this, a couple of things aren't clear to me:

* When defining multiple C exe's in one stanza, we need to manually pick which object files belong to which exe. I suppose we should just pick all objects except for those listed in `(names ..)`?

* When specifying a library as a dependency, which archives should we link? What makes sense is to use the `.a` file for the library. But this doesn't seem like what @avsm had in mind.

**Comment by diml** [comment]
> When defining multiple C exe's in one stanza, we need to manually pick which object files belong to which exe. I suppose we should just pick all objects except for those listed in (names ..)?

Yh, I don't have a better suggestion.

> When specifying a library as a dependency, which archives should we link? What makes sense is to use the .a file for the library. But this doesn't seem like what @avsm had in mind.

Maybe we should just not provide the `libraries` field for C executables? Personally, I would expect to have C library names in there, not OCaml library names,

**Comment by rgrinberg** [comment]
> Maybe we should just not provide the `libraries` field for C executables? Personally, I would expect to have C library names in there, not OCaml library names,

That's a good point but the problem we're trying to solve is to have headers attached to an OCaml library available when building a C executable. We wouldn't address the use case then.

However, it does seem a bit problematic to have a C executable without a corresponding library stanza. It also doesn't seem quite right that we have public header files that can't really be used by dune.

**Comment by diml** [comment]
I agree. Could you remind me what the use case is BTW?

**Comment by rgrinberg** [comment]
The problem is that we'd like to attach an `.h` file to a library and be able to compile against it whether the library is external or internal.

However, we decided that it was too low level when compared to dune's other features.

Looking at this again, it does not make sense to attach a `.h` to an OCaml library if you don't want the OCaml runtime. If the `.h` file needs to be standalone then it should be a C library not an OCaml one.

**Comment by diml** [comment]
Yeah, I suppose a `.h` file without corresponding object files is not super useful. That said, we could have OCaml libraries and C libraries live in the same namespace. Then it wouldn't be an issue to have a library name to which only `.h` files and/or C code not using the OCaml runtime are attached.

**Comment by rgrinberg** [comment]
I think that would work fine as long as we have a way to resolve the name conflicts. I think it's fairly common for OCaml libs to be named after their bindings so an OCaml `ssl` lib might shadow a C `ssl` lib. It seems to be undesirable to make this a conflict.

cc'ing @avsm here as some more motivation/description of the use case is necessary.

**Comment by diml** [comment]
We should be able to make that work relatively easily though as the two worls are separated. I.e. we can't have an external C library depend on an OCaml library.

For externel C library, I suggest that we stick to discovering them via configurator and just wrap them as OCaml libraries. In the future, we could add better support for pkg-config libraries, though we should probably not do that right now.

**Comment by avsm** [comment]
Am trying this patch out with the ctypes usecase and running into this error immediately:

```
$ dune build tests/test-constants/stub-generator/generated_stubs.exe
File "tests/test-constants/stub-generator/dune", line 15, characters 8-23:
15 |   (name generated_stubs)
             ^^^^^^^^^^^^^^^
Error: File unavailable: gcc
```

Can recreate this by cloning https://github.com/avsm/ocaml-ctypes/tree/dune-port and running:

```
dune build tests/test-constants/stub-generator/generated_stubs.exe
```

the relevant dune file is

```
(executable
  (name driver)
  (modules driver)
  (libraries ctypes test_constants_stubs tests_common))

(rule
  (targets generated_stubs.c)
  (action (run %{exe:driver.exe} --c-file %{targets})))

(rule
  (targets generated_bindings.ml)
  (action (run %{exe:driver.exe} --ml-file %{targets})))

(c_executable
  (name generated_stubs))
```

(the idea here is that the C file is generated by the Ctypes stub generator, which is then compiled with cc to generate the ml file that is then compiled as normal. 

I do have `/usr/bin/gcc` on the mac where this test is failing, but its actually llvm under the hood in case that makes a difference.

**Comment by rgrinberg** [comment]
@avsm note that the PR isn't in working state yet. The problem is that I'm not entirely sure how to proceed. The only to make sense of listing a library in the `libraries` field is to use the `.a` archive of that library. However, that would indeed include the OCaml runtime in the executable.

The only way to make sense of this feature is to have a way to define proper C libraries. Then you could specify exactly which object files you want to correspond to the installed header file.

**Comment by avsm** [comment]
I think we do need to be able to define `c_libraries` to use this sensibly.  The reason is that groups of object files will often need to be gathered into an archive and locally linked -- for examine see my [zarith/dune port](https://github.com/dune-universe/Zarith/commit/4b768894aeec7c8626dc66be2a40a482e24a43d6) or the vendored directory in [ocaml-yaml](https://github.com/avsm/ocaml-yaml).

On the bright side, this might replace the support for 'self building' C stubs that currently require manual rules in the `(library)` stanza.

**Comment by avsm** [comment]
Another thing I noticed in the Zarith port is that we could use ASFLAGS support at the same time as the C object file compilation, in order to support compiling `.S` files.  We have those in Zarith, and no doubt in crypto backends soon.

**Comment by gasche** [comment]
I needed something similar today and report in #10725 on what seems to work to build a simple C binary within a Dune project.

**Pull request closed without merging**